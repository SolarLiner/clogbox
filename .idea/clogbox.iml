<?xml version="1.0" encoding="UTF-8"?>
<module type="EMPTY_MODULE" version="4">
  <component name="FacetManager">
    <facet type="Python" name="Python facet">
      <configuration sdkName="Poetry (valib)" />
    </facet>
  </component>
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-core/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-math/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-clap/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-derive/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-filters/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-derive/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-graph/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/examples/svfmixer/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-enum/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-params/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-schedule/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/crates/clogbox-module/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/target" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>