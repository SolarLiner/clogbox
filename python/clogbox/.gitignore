.DS_Store
__pycache__
/junit
.coverage
coverage.xml