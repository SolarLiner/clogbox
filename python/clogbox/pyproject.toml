[project]
name = "clogbox"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
authors = [
    { name = "Nathan Graule", email = "solarliner@gmail.com" }
]
requires-python = ">=3.11"
dependencies = [
    "sympy>=1.14.0",
]

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[dependency-groups]
dev = [
    "pytest>=8.3.5",
    "syrupy>=4.9.1"
]
