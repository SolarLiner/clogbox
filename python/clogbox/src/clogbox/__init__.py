def main() -> None:
    print("Hello from clogbox!")
