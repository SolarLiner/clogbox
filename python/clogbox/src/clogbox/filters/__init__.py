import typing

import sympy as sp


class IntegratorInput(typing.NamedTuple):
    s: sp.Basic
    g: sp.Basic
    x: sp.Basic

Integrator: typing.TypeAlias = typing.Callable[[IntegratorInput], sp.Basic]


def linear_integrator(inp: IntegratorInput) -> sp.Basic:
    return inp.g * inp.x + inp.s


def ota_integrator(drive: sp.Basic, tanh=sp.tanh) -> typing.Callable[[IntegratorInput], sp.Basic]:
    return lambda inp: inp.g * tanh(drive * inp.x) / drive + inp.s


Shaper: typing.TypeAlias = typing.Callable[[sp.Basic], sp.Basic]

def drive(shaper: Shaper, drive: sp.Basic) -> Shaper:
    return lambda x: shaper(drive * x) / drive


def bias(shaper: Shaper, bias: sp.Basic) -> Shaper:
    return lambda x: shaper(x + bias)
