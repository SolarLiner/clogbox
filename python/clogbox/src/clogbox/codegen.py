import io
import typing
from dataclasses import dataclass
from typing import Optional, Tuple, Union

import sympy as sp
from sympy.codegen.ast import value_const
from sympy.codegen import ast
from sympy.printing.codeprinter import requires
from sympy.printing.rust import RustCodePrinter
import sympy.matrices.expressions as mat
from sympy.utilities import codegen

from clogbox.nodes import Tuple

RUST_CODE_PREAMBLE = """
//! Generated source code with the `clogbox` Python package, to use with the `clogbox` project.
#![allow(unused_imports, dead_code, non_snake_case, non_camel_case)]
use num_traits::{Float, FloatConst};
use az::CastFrom;
"""

DIFFERENTIABLE_CODE_PREAMBLE = (
    RUST_CODE_PREAMBLE
    + """
use clogbox_math::root_eq;
use nalgebra as na;
"""
)


class ClogboxRustCodePrinter(RustCodePrinter):
    tab = " " * 4

    def __init__(self, settings=None):
        if settings is None:
            settings = {}
        user_functions = settings.pop("user_functions", {})
        user_functions.update({"Determinant": "determinant"})
        settings["user_functions"] = user_functions
        super().__init__(settings)

        self.uses: set[str] = set()
        self.type_params: set[str] = set()

    @requires(type_params={"Float"})
    def _print_Pow(self, expr):
        base_code = self._print(expr.base)
        exp = expr.exp
        if isinstance(exp, sp.Integer):
            return f"{base_code}.powi({int(exp)})"
        else:
            return super()._print_Pow(expr)

    @requires(type_params={"CastFrom<f64>"}, uses={"az::CastFrom"})
    def _print_Integer(self, expr, _type=False):
        return f"T::cast_from({float(expr)})"

    @requires(type_params={"CastFrom<f64>"}, uses={"az::CastFrom"})
    def _print_Float(self, expr, **kwargs):
        return f"T::cast_from({float(expr)})"

    def _print_Rational(self, expr):
        p, q = tuple(self._print(sp.Integer(x)) for x in (expr.p, expr.q))
        return f"{p}/{q}"

    @requires(type_params={"na::Scalar"}, uses={"nalgebra as na"})
    def _print_MatrixBase(self, mat: sp.MatrixBase):
        values = ", ".join(self._print(x) for x in mat)
        cdim = f"na::Const<{mat.cols}>"
        if mat.rows == 1:
            mtype = f"na::OVector::<T, {cdim}>"
        else:
            rdim = f"na::Const<{mat.rows}>"
            mtype = f"na::OMatrix::<T, {rdim}, {cdim}>"
        return f"{mtype}::new({values})"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_NaN(self, expr, _type=False):
        return "T::nan()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Exp1(self, expr, _type=False):
        return "T::E()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Pi(self, expr, _type=False):
        return "T::PI()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Infinity(self, expr, _type=False):
        return "T::INFINITY()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_NegativeInfinity(self, expr, _type=False):
        return "T::NEG_INFINITY()"

    def _print_Tuple(self, expr: Tuple):
        return "({values})".format(values=", ".join(self._print(x) for x in expr.args))

    def _print_FunctionDefinition(self, fdecl: ast.FunctionDefinition):
        self.type_params.clear()

        params = ", ".join(self._print_Variable(var, _type=True) for var in fdecl.parameters)
        rtype = self._print_Type(fdecl.return_type)
        body = self._indent_codestring(self._print(fdecl.body))
        typeparams = "<T{bounds}>".format(bounds=": " + " + ".join(self.type_params) if len(self.type_params) > 0 else "")
        return f"""fn {fdecl.name}{typeparams}({params}) -> {rtype} {{\n{body}\n}}"""

    def _print_Type(self, t: ast.Type):
        if isinstance(t, ast.ComplexBaseType):
            self.uses.add("nalgebra as na")
            return "na::Complex<T>"
        if isinstance(t, ast.SignedIntType):
            return "i64"
        if isinstance(t, ast.UnsignedIntType):
            return "u64"
        return "T"

    def _print_Declaration(self, decl: ast.Declaration, _type=False):
        if value_const in decl.variable.attrs:
            mut = ""
        else:
            mut = " mut"
        return self._get_statement(f"let{mut} {self._print_Variable(decl.variable, _type=_type)} = {self._print(decl.variable.value)}")

    def _print_Variable(self, x: ast.Variable, _type=False):
        if _type:
            return f"{x.symbol.name}: {self._print_Type(x.type)}"
        return x.symbol.name

    def _print_While(self, node: ast.While):
        condition = self._print(node.condition)
        body = self._indent_codestring(self._print(node.body))
        return f"while {condition} {{\n{body}\n}}"

    def _print_Return(self, node: ast.Return, is_last_statement=False):
        if is_last_statement:
            return self._print(getattr(node, "return"))
        return self._get_statement(f"return {self._print(getattr(node, 'return'))}")

    def _print_Determinant(self, expr: mat.Determinant):
        return self._print_Function(expr)

    def _indent_codestring(self, codestring):
        return '\n'.join([self.tab + line for line in codestring.split('\n')])


class ClogboxCodegen(codegen.RustCodeGen):
    printer: ClogboxRustCodePrinter
    def __init__(self, **kwargs):
        kwargs["printer"] = ClogboxRustCodePrinter()
        super().__init__(**kwargs)

    def get_prototype(self, routine: codegen.Routine):
        def print_arg(arg: codegen.InputArgument | codegen.InOutArgument):
            self.printer.type_params.clear()

            typ = get_type(arg.name)
            if typ.startswith("na::"):
                self.printer.uses.add("nalgebra as na")
                self.printer.type_params.add("na::Scalar")
            return f"{arg.name}: {typ}"

        replacement, results = sp.cse(result.expr for result in typing.cast(list[codegen.Result], routine.results))
        codeblock = ast.CodeBlock(*(ast.Assignment(lhs, rhs) for lhs, rhs in replacement), ast.Return(*results))
        _body = self.printer.indent_code(self.printer.doprint(codeblock))
        args = ", ".join(print_arg(arg) for arg in routine.arguments)
        typeparams = "<T{bounds}>".format(bounds=": " + " + ".join(self.printer.type_params) if len(
            self.printer.type_params) > 0 else "")
        return f"fn {routine.name}{typeparams}({args}) -> T"


def render_as_module(content, printer: Optional[ClogboxRustCodePrinter]=None) -> str:
    if not printer:
        printer = ClogboxRustCodePrinter()
    str = printer.doprint(content)
    imports = "\n".join(f"use {module};" for module in printer.uses)
    return f"{imports}\n\n{str}"


def codegen_module(routines, printer: Optional[ClogboxRustCodePrinter]=None, codegen: Optional[ClogboxCodegen]=None,
                   project: str = "clogbox") -> str:
    if not printer:
        printer = ClogboxRustCodePrinter()
    if not codegen:
        codegen = ClogboxCodegen(printer=printer)

    [(_, contents)] = codegen.write(routines, "")  # type: str
    imports = "\n".join(f"use {module};" for module in codegen.printer.uses)
    end_of_comment = contents.find("*/") + 3 # 2 chars + newline
    return contents[:end_of_comment] + "\n" + imports + "\n\n" + contents[end_of_comment:]


def get_type(e: sp.Basic):
    """
    Get the Rust type of a sympy expression
    :param e: Sympy expression
    :return: Rust type representation
    """
    match e:
        case sp.MatrixBase(cols=c, rows=r) | sp.MatrixSymbol(cols=c, rows=r):
            return f"na::OMatrix<T, na::Const<{r}>, na::Const<{c}>>"
        case _:
            return "T"


@dataclass()
class IndentStream(io.TextIOBase):
    stream: io.TextIOBase
    indent: str = " " * 4

    def write(self, s: str) -> int:
        return self.stream.write(s.replace("\n", "\n" + self.indent))

    def flush(self):
        self.stream.flush()

    def close(self):
        self.stream.close()


def generate_codeblock(exprs: list[sp.Expr]) -> ast.CodeBlock:
    replacements, exprs = sp.cse(exprs)
    return ast.CodeBlock(*(ast.Assignment(lhs, rhs) for lhs, rhs in replacements), Tuple(exprs))

def generate_differentiable(
    f: io.TextIOBase,
    expr: Union[sp.Expr, sp.Eq],
    variable: Union[sp.Symbol, sp.MatrixBase],
    struct_name="Equation",
    printer: Optional[RustCodePrinter] = None,
    codegen: Optional[ClogboxCodegen] = None,
    runtime_invert=False,
):
    """
    Generate Rust code for a specific equation that is differentiable once about a variable. This function will
    implement the `clogbox_math::root_eq::Differentiable` trait for the generated struct.

    The resulting source code needs the `clogbox_math` and `num-traits` crates. The final module must also have the
    contents of the `DIFFERENTIABLE_CODE_PREAMBLE` constant present.

    :param f: Stream to write the code to.
    :param expr: Expression to implement the trait for.
    :param variable: Variable about which the equation is differentiable.
    :param struct_name: Name of the struct to generate (defaults to "Equation").
    :param printer: Optional instance of `RustCodePrinter` used to print symbolic
        expressions as Rust-compatible code. If not provided, a default printer will
        be used.
    """
    if not printer:
        printer = ClogboxRustCodePrinter()
    if not codegen:
        codegen = ClogboxCodegen(printer=printer)

    if isinstance(expr, sp.Eq):
        expr = expr.rhs - expr.lhs

    if isinstance(variable, sp.MatrixBase):
        var_symbols = set(variable.free_symbols)
    else:
        var_symbols = {variable}
    fields = sorted(expr.free_symbols - var_symbols, key=lambda s: s.name)

    f.write(f"pub struct {struct_name}<T> {{")
    for field in fields:
        f.write(f"    pub {str(field)}: {get_type(field)},")
    f.write( "}")
    f.writelines([""])

    if isinstance(expr, sp.MatrixBase):
        vtype = "na::OVector<Self::Scalar, Self::Dim>"
        mtype = "na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>"
        vview_type = "na::VectorView<Self::Scalar, Self::Dim>"
        f.write(
f"""
impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for {struct_name}<T> {{
    type Scalar = T;
    type Dim = na::Const<{expr.rows}>;
    
    fn eval_with_inv_jacobian(&self, matrix: {vview_type}) -> ({vtype}, {mtype}) {{"""
        )
        for i, name in enumerate(variable):
            f.write(f"        let {str(name)} = matrix[{i}];")
        for field in fields:
            f.write(f"        let {str(field)} = self.{str(field)};")

        s = IndentStream(f, indent=" " * 8)
        if runtime_invert:
            j = expr.jacobian(variable)
            block = generate_codeblock([expr, j])
            s.write(printer.doprint(block, assign_to="(f, df)"))
            s.write("        (f, df.try_inverse().unwrap())")
        else:
            inv_j = expr.jacobian(variable).inverse()
            block = generate_codeblock([expr, inv_j])
            s.write(printer.doprint(block))

    else:
        f.write(
f"""
impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::Differentiable for {struct_name}<T> {{
    type Scalar = T;

    fn eval_with_derivative(&self, {str(variable)}: T) -> (T, T) {{
"""
        )
        for field in fields:
            f.write(f"        let {str(field)} = self.{str(field)};")

        diff = sp.diff(expr, variable)
        block = generate_codeblock([expr, diff])
        f.write(printer.doprint(block))
        f.write(
"""
    }
}
"""
        )