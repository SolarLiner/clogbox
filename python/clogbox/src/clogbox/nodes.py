from sympy.codegen import ast


class Tuple(ast.Node):
    __slots__ = ('elements', *ast.Node.__slots__)
    _fields = ('elements', *ast.Node._fields)

    def __new__(cls, *args, **kwargs):
        if len(args) == 1 and len(kwargs) == 0:
            return args[0]
        return ast.Node.__new__(cls, *args, **kwargs)
