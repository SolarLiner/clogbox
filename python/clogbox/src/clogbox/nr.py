import sympy as sp

from clogbox.codegen import ClogboxRustCodePrinter

NR_RUST_PREAMBLE = """
use clogbox_math::root_eq::Differentiable;
"""


def generate_root_eq(equation, variable, struct_name ="Equation"):
    """
    Generate Rust code for a specific equation to be solved with the Newton-Raphson method,
    optimized using common subexpression elimination.

    Parameters:
    -----------
    equation : sympy.Expr
        A sympy expression that will be solved for zero (i.e., equation = 0)
    variable : sympy.Symbol
        The variable in the equation

    Returns:
    --------
    str
        Rust source code implementing the specific equation using the Differentiable trait,
        including all free variables as fields in the struct and using generics with Float bound.
    """
    # "Normalize" an equality to a root equation
    if isinstance(equation, sp.Eq):
        equation = equation.rhs - equation.lhs

    # Calculate the derivative
    derivative = sp.diff(equation, variable)

    # Perform common subexpression elimination on both expressions
    replacements, reduced_exprs = sp.cse([equation, derivative])

    # Use the custom printer
    printer = ClogboxRustCodePrinter()

    # Generate Rust code for the common subexpressions
    cse_code = ""
    for idx, (sym, expr) in enumerate(replacements):
        cse_code += f"let {printer.doprint(sym)} = {printer.doprint(expr)};\n        "

    # Generate Rust code for the reduced expressions
    eq_str = printer.doprint(reduced_exprs[0])
    derivative_str = printer.doprint(reduced_exprs[1])

    # Identify all free symbols except the main variable
    free_vars = sorted(equation.free_symbols - {variable}, key=lambda s: s.name)

    # Create struct fields for each free variable
    struct_fields = "\n    ".join(f"pub {str(var)}: T," for var in free_vars)
    param_inserts = "\n        ".join(f"let {str(var)} = self.{str(var)};" for var in free_vars)

    x = variable.name
    # Generate the Rust code
    rust_code = f"""/// Structure representing the specific equation to solve
#[derive(Debug, Clone, Copy)]
pub struct {struct_name}<T> {{
    {struct_fields}
}}

impl<T: Float + FloatConst + CastFrom<f64>> Differentiable for {struct_name}<T> {{
    type Scalar = T;

    fn eval(&self, {x}: T) -> T {{
        {param_inserts}
        {cse_code}
        {eq_str}
    }}

    fn derivative(&self, {x}: T) -> T {{
        {param_inserts}
        {cse_code}
        {derivative_str}
    }}

    fn eval_with_derivative(&self, {x}: T) -> (T, T) {{
        {param_inserts}
        {cse_code}
        let fx = {eq_str};
        let dfx = {derivative_str};
        (fx, dfx)
    }}
}}
"""

    return rust_code
