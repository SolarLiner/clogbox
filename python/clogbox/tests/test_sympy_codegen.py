from io import StringIO

import sympy as sp

from clogbox.codegen import ClogboxRustCodePrinter, ClogboxCodegen, render_as_module, codegen_module, generate_differentiable

import pytest


@pytest.fixture
def codegen(printer: ClogboxRustCodePrinter) -> ClogboxCodegen:
    return ClogboxCodegen(printer=printer)


def test_abs_diff(printer: ClogboxRustCodePrinter, snapshot: str):
    x = sp.Symbol("x", real=True)
    expr = sp.Abs(x).diff(x)
    actual = printer.doprint(expr)
    assert snapshot == actual


def test_hyperbolic_diff(printer: ClogboxRustCodePrinter, snapshot: str):
    x = sp.Symbol("x", real=True)
    hyperbolic = x / (1 + sp.Abs(x))
    actual = StringIO()
    generate_differentiable(actual, hyperbolic, x, printer=printer)
    assert snapshot == actual.getvalue()


def test_piecewise(printer: ClogboxRustCodePrinter, snapshot: str):
    x = sp.Symbol("x", real=True)
    expr = sp.Piecewise((x, x < 0), (x ** 2, True))
    actual = printer.doprint(expr)
    assert snapshot == actual


def test_asinh_derivative(printer: ClogboxRustCodePrinter, snapshot: str):
    u = sp.Symbol('u', real=True)
    expr = 1 / sp.sqrt(u ** 2 + 1)
    actual = printer.doprint(expr)
    # expected = "(u.powi(2) + T::cast_from(1.0)).sqrt().recip()"
    assert snapshot == actual


def test_generate_differentiable(codegen: ClogboxCodegen, snapshot: str):
    u = sp.Symbol('u', real=True)
    expr = sp.asinh(u) - sp.tanh(u)

    actual = StringIO()
    generate_differentiable(actual, expr, u, codegen=codegen)
    assert snapshot == actual.getvalue()


def test_newton_rhapson_function(printer: ClogboxRustCodePrinter, snapshot: str):
    from sympy.codegen.algorithms import newtons_method_function

    u = sp.Symbol('u', real=True)
    expr = sp.asinh(u) - sp.tanh(u)
    code = newtons_method_function(expr, u, cse=True)
    actual = printer.doprint(code)
    assert snapshot == actual


def test_matrix_expression(printer: ClogboxRustCodePrinter, snapshot: str):
    u = sp.Symbol("u")
    X = sp.Matrix([[u, u ** 2], [u ** 3, u ** 4]])
    root_eq = sp.Determinant(X) - sp.tanh(u)
    actual = printer.doprint(root_eq)
    assert snapshot == actual


def test_matrix_routine(codegen: ClogboxCodegen, snapshot: str):
    X = sp.MatrixSymbol('X', 2, 2)
    y = sum(X) / sp.Determinant(X)

    routine = codegen.routine("matrix_routine", y, [X], [])
    actual = codegen_module([routine], codegen=codegen)
    assert snapshot == actual
