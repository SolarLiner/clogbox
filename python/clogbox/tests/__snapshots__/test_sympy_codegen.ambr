# serializer version: 1
# name: test_abs_diff
  '(x).signum()'
# ---
# name: test_asinh_derivative
  '(u.powi(2) + T::cast_from(1.0)).sqrt().recip()'
# ---
# name: test_generate_differentiable
  '''
  pub struct Equation<T> {
  }
  
  impl<T: Copy + Float + FloatConst + CastFrom<f64>> root_eq::Differentiable for Equation<T> {
      type Scalar = T;
  
      fn eval_with_derivative(&self, u: T) -> (T, T) {
          {
              let x0 = u.tanh();
              (-x0 + u.asinh(), x0.powi(2) + T::cast_from(-1.0) + (u.powi(2) + T::cast_from(1.0)).sqrt().recip())
          }
      }
  }
  '''
# ---
# name: test_hyperbolic_diff
  '''
  pub struct Equation<T> {
  }
  
  impl<T: Copy + Float + FloatConst + CastFrom<f64>> root_eq::Differentiable for Equation<T> {
      type Scalar = T;
  
      fn eval_with_derivative(&self, x: T) -> (T, T) {
          {
              let x0 = x.abs() + T::cast_from(1.0);
              let x1 = x0.powi(-1);
              (x*x1, -x*(x).signum()/x0.powi(2) + x1)
          }
      }
  }
  '''
# ---
# name: test_matrix_expression
  '-u.tanh() + (na::OMatrix::<T, na::Const<2>, na::Const<2>>::new(u, u.powi(2), u.powi(3), u.powi(4))).determinant()'
# ---
# name: test_matrix_routine
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use nalgebra as na;
  use num_traits::Float;
  
  
  pub fn matrix_routine<T: Float + na::Scalar>(X: na::OMatrix<T, na::Const<2>, na::Const<2>>) -> T {
  
      let out1 = (X[0] + X[1] + X[2] + X[3])/(X).determinant();
      out1
  
  }
  
  '''
# ---
# name: test_newton_rhapson_function
  '''
  pub fn newton<T: CastFrom<f64> + Float + FloatConst>(u: T) -> T {
      let mut d_u = T::INFINITY();
      while d_u.abs() > T::cast_from(4e-16)*u.abs() + T::cast_from(1e-12) {
          x0 = (u.powi(2) + T::cast_from(1.0)).sqrt();
          x1 = u.tanh();
          d_u = -x0*(-x1 + u.asinh())/(x0*x1.powi(2) - x0 + T::cast_from(1.0));
          u += d_u;
      }
      u
  }
  '''
# ---
# name: test_piecewise
  '''
  match () {
      _ if x < T::zero() => x,
      _ => x.powi(2)
  }
  '''
# ---
